x/tokenfactory/keeper/admins_test.go

```diff
@@ -1,74 +1,270 @@
+			if tc.valid {
+				addr0bal += 10
+				suite.Require().NoError(err)
+				suite.Require().Equal(suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom).Amount.Int64(), addr0bal, suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom))
+			} else {
+				suite.Require().Error(err)
+			}
+		})
+	}
+}
+
+func (suite *KeeperTestSuite) TestBurnDenom() {
+	var (
+		addr0bal = int64(0)
+	)
+
+	// Create a denom and mint
+	suite.CreateDefaultDenom()
+
+	// Make sure that the admin is set correctly
+	queryRes, err := suite.queryClient.DenomAuthorityMetadata(suite.Ctx.Context(), &types.QueryDenomAuthorityMetadataRequest{
+		Denom: suite.defaultDenom,
+	})
+	suite.Require().NoError(err)
+	suite.Require().Equal(suite.TestAccs[0].String(), queryRes.AuthorityMetadata.Admin)
+
+	_, err = suite.msgServer.Mint(sdk.WrapSDKContext(suite.Ctx), types.NewMsgMint(suite.TestAccs[0].String(), sdk.NewInt64Coin(suite.defaultDenom, 10)))
+	addr0bal += 10
+	suite.Require().NoError(err)
+	suite.Require().True(suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom).Amount.Int64() == addr0bal, suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom))
+
+	for _, tc := range []struct {
+		desc      string
+		amount    int64
+		burnDenom string
+		admin     string
+		valid     bool
+	}{
+		{
+			desc:      "denom does not exist",
+			amount:    10,
+			burnDenom: "factory/osmo1t7egva48prqmzl59x5ngv4zx0dtrwewc9m7z44/evmos",
+			admin:     suite.TestAccs[0].String(),
+			valid:     false,
+		},
+		{
+			desc:      "burn is not by the admin",
+			amount:    10,
+			burnDenom: suite.defaultDenom,
+			admin:     suite.TestAccs[1].String(),
+			valid:     false,
+		},
+		{
+			desc:      "burn amount is bigger than minted amount",
+			amount:    1000,
+			burnDenom: suite.defaultDenom,
+			admin:     suite.TestAccs[1].String(),
+			valid:     false,
+		},
+		{
+			desc:      "success case",
+			amount:    10,
+			burnDenom: suite.defaultDenom,
+			admin:     suite.TestAccs[0].String(),
+			valid:     true,
+		},
+	} {
+		suite.Run(fmt.Sprintf("Case %s", tc.desc), func() {
+			// Test minting to admins own account
+			_, err := suite.msgServer.Burn(sdk.WrapSDKContext(suite.Ctx), types.NewMsgBurn(tc.admin, sdk.NewInt64Coin(tc.burnDenom, 10)))
+			if tc.valid {
+				addr0bal -= 10
+				suite.Require().NoError(err)
+				suite.Require().True(suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom).Amount.Int64() == addr0bal, suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom))
+			} else {
+				suite.Require().Error(err)
+				suite.Require().True(suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom).Amount.Int64() == addr0bal, suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom))
+			}
+		})
+	}
+}
+
+func (suite *KeeperTestSuite) TestChangeAdminDenom() {
```

Comment:
This tests logic due to the complex malleate's isn't very easy to follow. I'll spend some more time tmrw thinking about how to restructure it

Reply:
What if instead of setup(), we added three fields:
* `ChangeAdminMsg`
* `exoectedAdmin`
* `*MintMsg` (and can be nil)
* `expMintPass`

Sorry for the delay it took me to come back to this